File: src/dim_tuning_utils.c
/****************************************************************************
 *                           Dim tuning utilities                           *
 ****************************************************************************/
#include "dim_tuning_utils.h"


/****************************************************************************
 * Dim tuning and the 'dim_tuner' argument
 *
 * Dim tuning
 * ----------
 * Dim tuning is the act of adding and/or dropping ineffective dimensions,
 * (i.e. dimensions that have an extent of 1) to/from an array-like object.
 * Note that dim tuning doesn't change the length (which is prod(dim(.)))
 * or alter the content of the object, and is always reversible (except when
 * it drops ineffective dimensions with dimnames on them).
 *
 * The 'dim_tuner' argument
 * ------------------------
 * The exact action to perform on the dimensions of the object is encoded
 * in the 'dim_tuner' argument. This is an integer vector where each value
 * represents one of three possible operations:
 *   o  0: Keep the dimension.
 *   o -1: Drop the dimension. This operation is allowed only if the
 *         dimension to drop is ineffective (i.e. has an extent of 1).
 *   o  1: Add ineffective dimension.
 * Note that the 'dim_tuner' vector can contain any number of 1's, but the
 * number of non-positive values (i.e. number of 0 and -1 values together)
 * must match the number of dimensions of the array-like object to tune.
 *
 * Additionally, 'dim_tuner' must contain at least one 0. In other words,
 * the tuning must retain at least one of the original dimensions of the
 * object.
 *
 * Reverse tuning
 * --------------
 * To revert a dim tuning, simply tune again with '-dim_tuner' (i.e. minus
 * 'dim_tuner'). More precisely, for S4Arrays:::tune_dims() this means that
 * 'dim2' will always be identical to 'dim' here:
 *
 *      tuned_dim <- S4Arrays:::tune_dims(dim, dim_tuner)
 *      dim2 <- S4Arrays:::tune_dims(tuned_dim, -dim_tuner)
 *
 * This should be TRUE for any vector of dimensions 'dim' and any 'dim_tuner'
 * vector compatible with 'dim'.
 */

#define	KEEP_DIM        0
#define	DROP_DIM       -1
#define	ADD_DIM         1

/* Return the "new" number of dimensions i.e. the number of dims that we
   will get after tuning the current vector of dimensions. Note that this
   is simply the number of non-negative values in 'ops' (i.e. number of
   0 and 1 values together). */
static int validate_dim_tuner(const int *ops, int nops,
		const int *dims, int ndim)
{
	int along1, along2, nkept, r, op;

	along1 = along2 = nkept = 0;
	for (r = 0; r < nops; r++) {
		op = ops[r];  /* ADD_DIM, KEEP_DIM, or DROP_DIM */
		if (op == ADD_DIM) {
			along2++;
			continue;
		}
		if (along1 >= ndim)
			error("S4Arrays internal error in "
			      "validate_dim_tuner():\n"
			      "    number of 0 (KEEP) or -1 (DROP) values "
			      "in 'dim_tuner' is > 'length(dim(x))'");
		if (op == KEEP_DIM) {
			along2++;
			nkept++;
			along1++;
			continue;
		}
		if (op != DROP_DIM)
			error("S4Arrays internal error in "
			      "validate_dim_tuner():\n"
			      "    'dim_tuner' can only contain 0 (KEEP), "
			      "-1 (DROP), or 1 (ADD) values");
		if (dims[along1] != 1)
			error("S4Arrays internal error in "
			      "validate_dim_tuner():\n"
			      "    'dim_tuner[%d]' (= -1) is "
			      "mapped to 'dim(x)[%d]' (= %d)\n"
			      "    which cannot be dropped",
			      r + 1, along1 + 1, dims[along1]);
		along1++;
	}
	if (along1 < ndim)
		error("S4Arrays internal error in "
		      "validate_dim_tuner():\n"
		      "    number of 0 (KEEP) or -1 (DROP) values "
		      "in 'dim_tuner' is < 'length(dim(x))'");
	if (nkept == 0)
		error("S4Arrays internal error in "
		      "validate_dim_tuner():\n"
		      "    'dim_tuner' must contain at least one 0");
	return along2;
}


/****************************************************************************
 * C_tune_dims() and C_tune_dimnames()
 */

/* 'dim_names' must be a character vector (or R_NilValue) that contains the
   names on the vector of dimensions. This is NOT the same as the dimnames! */
static SEXP tune_dims(const int *dims, SEXP dim_names,
		const int *ops, int nops, int ans_len)
{
	SEXP ans, ans_names = R_NilValue;
	int along1, along2, r, op;

	ans = PROTECT(NEW_INTEGER(ans_len));
	if (dim_names != R_NilValue)
		ans_names = PROTECT(NEW_CHARACTER(ans_len));
	along1 = along2 = 0;
	for (r = 0; r < nops; r++) {
		op = ops[r];  /* ADD_DIM, KEEP_DIM, or DROP_DIM */
		if (op == ADD_DIM) {
			INTEGER(ans)[along2] = 1;
			along2++;
			continue;
		}
		if (op == KEEP_DIM) {
			INTEGER(ans)[along2] = dims[along1];
			if (dim_names != R_NilValue)
				SET_STRING_ELT(ans_names, along2,
					       STRING_ELT(dim_names, along1));
			along2++;
		}
		along1++;
	}
	if (dim_names != R_NilValue) {
		SET_NAMES(ans, ans_names);
		UNPROTECT(1);
	}
	UNPROTECT(1);
	return ans;
}

static SEXP tune_dimnames(SEXP dimnames,
		const int *ops, int nops, int ans_len)
{
	int along1, along2, r, op;
	SEXP ans;

	ans = PROTECT(NEW_LIST(ans_len));
	along1 = along2 = 0;
	for (r = 0; r < nops; r++) {
		op = ops[r];  /* ADD_DIM, KEEP_DIM, or DROP_DIM */
		if (op == ADD_DIM) {
			along2++;
			continue;
		}
		if (op == KEEP_DIM) {
			SET_VECTOR_ELT(ans, along2,
				       VECTOR_ELT(dimnames, along1));
			along2++;
		}
		along1++;
	}
	UNPROTECT(1);
	return ans;
}

/* Return the length of the tuned 'dimnames' or 0 if the tuning retains no
   dimnames. */
static int compute_tuned_dimnames_length(SEXP dimnames,
		const int *ops, int nops)
{
	int ndim, along1, along2, any_retained, r, op;

	if (dimnames == R_NilValue)
		return 0;
	ndim = LENGTH(dimnames);
	along1 = along2 = any_retained = 0;
	for (r = 0; r < nops; r++) {
		op = ops[r];  /* ADD_DIM, KEEP_DIM, or DROP_DIM */
		if (op == ADD_DIM) {
			along2++;
			continue;
		}
		if (along1 >= ndim)
			error("S4Arrays internal error in "
			      "compute_tuned_dimnames_length():\n"
			      "    number of 0 (KEEP) or -1 (DROP) values "
			      "in 'dim_tuner' is > 'length(dim(x))'");
		if (op == KEEP_DIM) {
			if (VECTOR_ELT(dimnames, along1) != R_NilValue)
				any_retained = 1;
			along2++;
		}
		along1++;
	}
	return any_retained ? along2 : 0;
}

/* --- .Call ENTRY POINT --- */
SEXP C_tune_dims(SEXP dim, SEXP dim_tuner)
{
	int ndim, nops, ans_len;
	const int *dims, *ops;

	ndim = LENGTH(dim);
	dims = INTEGER(dim);
	nops = LENGTH(dim_tuner);
	ops = INTEGER(dim_tuner);
	ans_len = validate_dim_tuner(ops, nops, dims, ndim);
	return tune_dims(dims, GET_NAMES(dim), ops, nops, ans_len);
}

/* --- .Call ENTRY POINT --- */
SEXP C_tune_dimnames(SEXP dimnames, SEXP dim_tuner)
{
	int nops, ans_len;
	const int *ops;

	nops = LENGTH(dim_tuner);
	ops = INTEGER(dim_tuner);
	ans_len = compute_tuned_dimnames_length(dimnames, ops, nops);
	if (ans_len == 0)
		return R_NilValue;
	return tune_dimnames(dimnames, ops, nops, ans_len);
}

File: R/dim-tuning-utils.R
### =========================================================================
### Dim tuning utilities
### -------------------------------------------------------------------------
###
### "Dim tuning" is the act of adding and/or dropping "ineffective
### dimensions" to/from an array-like object, typically via the drop()
### and/or dim() setter. The exact transformation to operate on the vector
### of dimensions of the object can be precisely described by supplying
### a 'dim_tuner' vector.
### See src/dim_tuning_utils.c for additional information.


### NOT exported but used in the DelayedArray packages!
normalize_dim_replacement_value <- function(value, x_dim)
{
    if (is.null(value))
        stop(wmsg("you can't do that, sorry"))
    if (!is.numeric(value))
        stop(wmsg("the supplied dim vector must be numeric"))
    if (length(value) == 0L)
        stop(wmsg("the supplied dim vector cannot be empty"))
    if (!is.integer(value))
        value <- as.integer(value)
    if (S4Vectors:::anyMissingOrOutside(value, 0L))
        stop(wmsg("the supplied dim vector cannot contain negative ",
                  "or NA values"))
    prod1 <- prod(value)
    prod2 <- prod(x_dim)
    if (prod1 != prod2)
        stop(wmsg("the supplied dims [product ", prod1, "] do not match ",
                  "the length of object [", prod2, "]"))
    value
}


### - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
### The tune_Array_dims() low-level generic
###
### Array derivatives (e.g. SparseArray or DelayedArray objects) only need
### to implement a tune_Array_dims() method to have drop() and the dim()
### setter work out-of-the-box.
###
### Note that a "dim tuning" operation does NOT change the length of the
### object (which is prod(dim(x))) or alter its content, and should always
### be reversible (except when it drops ineffective dimensions with dimnames
### on them). To revert a "dim tuning" operation, simply tune again
### with '-dim_tuner' (i.e. with minus 'dim_tuner'). More precisely, for
### tune_Array_dims(), 'x2' should always be identical to 'x' here:
###
###     y <- tune_Array_dims(x, dim_tuner)
###     x2 <- tune_Array_dims(y, -dim_tuner)
###     identical(x2, x)  # should be TRUE
###
### This should be the case for any array-like object 'x' (with no dimnames
### on its ineffective dimensions) and any 'dim_tuner' vector compatible
### with 'dim(x)'.

setGeneric("tune_Array_dims", signature="x",
    function(x, dim_tuner) standardGeneric("tune_Array_dims")
)


### - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
### tune_dims() and tune_dimnames()
###
### NOT exported but used by the tune_Array_dims() method for SVT_SparseArray
### objects defined in the SparseArray package.

tune_dims <- function(dim, dim_tuner)
{
    stopifnot(is.integer(dim),
              is.integer(dim_tuner))
    .Call2("C_tune_dims", dim, dim_tuner, PACKAGE="S4Arrays")
}

tune_dimnames <- function(dimnames, dim_tuner)
{
    stopifnot(is.null(dimnames) || is.list(dimnames),
              is.integer(dim_tuner))
    .Call2("C_tune_dimnames", dimnames, dim_tuner, PACKAGE="S4Arrays")
}


### - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
### drop() method
###

### Expected to be semantically equivalent to 'drop(as.array(x))'.
### Will work out-of-the-box on any Array derivative that supports
### tune_Array_dims() and as.array(). Note that the latter is used
### only if 'x' has at most one effective dimension.
.drop_Array <- function(x)
{
    is_effective <- dim(x) != 1L
    if (sum(is_effective) <= 1L)
        return(drop(as.array(x)))  # ordinary vector
    dim_tuner <- -as.integer(!is_effective)
    tune_Array_dims(x, dim_tuner)
}

setMethod("drop", "Array", .drop_Array)


### - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
### dim() setter
###

.diff_dims <- function(old_dim, new_dim, x_class)
{
    stopifnot(is.integer(old_dim), is.integer(new_dim))

    cannot_map_msg <- c(
        "Cannot map the supplied dim vector to the current dimensions of ",
        "the object. On a ", x_class, " object, the dim() setter can only ",
        "be used to drop and/or add \"ineffective dimensions\" (i.e. ",
        "dimensions equal to 1) to the object."
    )
    can_map_effective_dimensions <- function(effdim_idx1, effdim_idx2) {
        if (length(effdim_idx1) != length(effdim_idx2))
            return(FALSE)
        if (length(effdim_idx1) == 0L)
            return(TRUE)
        all(old_dim[effdim_idx1] == new_dim[effdim_idx2])
    }

    ## Get index of old and new effective dimensions.
    effdim_idx1 <- which(old_dim != 1L)
    effdim_idx2 <- which(new_dim != 1L)

    if (!can_map_effective_dimensions(effdim_idx1, effdim_idx2))
        stop(wmsg(cannot_map_msg))

    compute_dim_tuner <- function(effdim_idx1, effdim_idx2) {
        idx1 <- c(effdim_idx1, length(old_dim) + 1L)
        idx2 <- c(effdim_idx2, length(new_dim) + 1L)
        diffs1 <- S4Vectors:::diffWithInitialZero(idx1)
        diffs2 <- S4Vectors:::diffWithInitialZero(idx2)
        deltas <- pmax(diffs1, diffs2)
        nonzero_runlengths <- deltas - 1L
        ans_len <- sum(nonzero_runlengths) + length(effdim_idx1)
        ans <- integer(ans_len)
        offsets <- c(0L, head(cumsum(deltas), n=-1L))
        for (k in seq_along(idx1)) {
            d <- diffs2[[k]] - diffs1[[k]]
            op <- ifelse(d > 0L, 1L, -1L)
            ans[offsets[[k]] + seq_len(abs(d))] <- rep.int(op, abs(d))
        }
        ans
    }

    compute_dim_tuner(effdim_idx1, effdim_idx2)
}

.set_Array_dim <- function(x, value)
{
    x_dim <- dim(x)
    value <- normalize_dim_replacement_value(value, x_dim)
    dim_tuner <- .diff_dims(x_dim, value, class(x))
    ans <- tune_Array_dims(x, dim_tuner)
    stopifnot(identical(dim(ans), value))  # sanity check
    ans
}

setReplaceMethod("dim", "Array", .set_Array_dim)

